runtime-modules/project-token/src/types.rs

```diff
@@ -0,0 +1,414 @@
+            self.unclaimed_patronage_at_block(block);
+        self.patronage_info.last_unclaimed_patronage_tally_block = block;
+        self.patronage_info.rate = new_rate;
+    }
+}
+
+/// Encapsules parameters validation + TokenData construction
+impl<Balance: Zero + Copy + PartialOrd, Hash> TokenIssuanceParameters<Balance, Hash> {
+    /// Forward `self` state
+    pub fn build<BlockNumber, T: crate::Trait>(
+        self,
+        block: BlockNumber,
+    ) -> TokenData<Balance, Hash, BlockNumber> {
+        // validation
+        let rate = if self.patronage_rate.0.is_zero() {
+            BlockRate(Perquintill::zero())
+        } else {
+            BlockRate::from_yearly_rate(self.patronage_rate, T::BlocksPerYear::get())
+        };
+        let patronage_info = PatronageData::<Balance, BlockNumber> {
+            last_unclaimed_patronage_tally_block: block,
+            unclaimed_patronage_tally_amount: Balance::zero(),
+            rate,
+        };
+        TokenData::<Balance, Hash, BlockNumber> {
+            supply: self.initial_supply,
+            offering_state: OfferingState::Idle,
+            transfer_policy: self.transfer_policy,
+            symbol: self.symbol,
+            patronage_info,
+            accounts_number: 0u64,
+        }
+    }
+}
+
+impl<Hasher: Hash> MerkleProof<Hasher> {
+    pub(crate) fn verify_for_commit<T, AccountId>(
+        &self,
+        account_id: &AccountId,
+        commit: Hasher::Output,
+    ) -> DispatchResult
+    where
+        T: crate::Trait,
+        AccountId: Encode,
+    {
+        let init = Hasher::hash_of(account_id);
+        let proof_result = self.0.iter().fold(init, |acc, (hash, side)| match side {
+            MerkleSide::Left => Hasher::hash_of(&(hash, acc)),
+            MerkleSide::Right => Hasher::hash_of(&(acc, hash)),
+        });
+
+        ensure!(
+            proof_result == commit,
+            crate::Error::<T>::MerkleProofVerificationFailure,
+        );
+
+        Ok(())
+    }
+}
+
+impl<AccountId, Balance: Sum + Copy> Transfers<AccountId, Balance> {
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn total_amount(&self) -> Balance {
+        self.0.iter().map(|(_, payment)| payment.amount).sum()
+    }
+
+    pub fn iter(&self) -> Iter<'_, AccountId, Payment<Balance>> {
+        self.0.iter()
+    }
+
+    pub fn into_iter(self) -> IntoIter<AccountId, Payment<Balance>> {
+        self.0.into_iter()
+    }
+}
+
+impl<AccountId, Balance> From<Transfers<AccountId, Balance>>
+    for BTreeMap<AccountId, Payment<Balance>>
+{
+    fn from(v: Transfers<AccountId, Balance>) -> Self {
+        v.0
+    }
+}
+
+/// Block Rate bare minimum impementation
+impl BlockRate {
+    pub fn from_yearly_rate(r: YearlyRate, blocks_per_year: u32) -> Self {
+        BlockRate(Perquintill::from_rational_approximation(
+            1u64,
+            r.0.saturating_reciprocal_mul(blocks_per_year as u64),
```

Comment:
I'm not sure what happens if the `YearlyRate` is `0`, since `r.0.saturating_reciprocal_mul(blocks_per_year as u64)` is _Inifinity_ in that case. Probably worth adding a test case and securing against if needed.